week3/uchan0/app.js

```diff
@@ -0,0 +1,141 @@
+
+    listEl.append(deleteEl);
+    listEl.append(updateEl);
+    todoListEl.append(listEl);
+  });
+};
+
+const addTodo = () => {
+  const title = todoInputEl.value;
+  const date = new Date();
+  const createdAt = date.toDateString();
+  if (!title) return;
+
+  const newTodo = {
+    id: date.getTime().toString(),
+    title,
+    createdAt,
+  };
+
+  fetch(API_URL, {
+    method: "POST",
+    headers: {
+      "Content-Type": "application/json",
+    },
+    body: JSON.stringify({ ...newTodo, completed: false }),
+  })
+    .then((response) => response.json())
+    .then(() => {
+      // 최신 목록을 다시 가져와서 렌더링
+      todoInputEl.value = "";
+      return fetch(API_URL);
+    })
+    .then((response) => response.json().then((data) => renderTodo(data)));
+};
+
+const deleteTodo = (todoId) => {
+  fetch(API_URL + "/" + todoId, {
+    method: "DELETE",
+  })
+    // 최신 목록을 다시 가져와서 렌더링
+    .then(() => fetch(API_URL))
+    .then((response) => response.json())
+    .then((data) => renderTodo(data));
+};
+
+//mission
+const updateTodo = (todoId, originalTitle) => {
+  const todoItem = document.querySelector(`#todo-${todoId}`);
+
+  //form 태그 생성
+  const formEl = document.createElement("form");
+
+  // 입력창 생성
+  const inputEl = document.createElement("input");
+  inputEl.type = "text";
+  inputEl.value = "";
+
+  // 확인버튼 생성
+  const confirmBtn = document.createElement("button");
+  confirmBtn.textContent = "확인";
+  confirmBtn.type = "submit";
+
+  // 취소버튼 생성
+  const cancelBtn = document.createElement("button");
+  cancelBtn.textContent = "취소";
+  cancelBtn.type = "button";
+  cancelBtn.style.backgroundColor = "lightgray";
+
+  // form태그 안에 input, 버튼 추가
+  formEl.appendChild(inputEl);
+  formEl.appendChild(confirmBtn);
+  formEl.appendChild(cancelBtn);
+
+  todoItem.innerHTML = "";
+  todoItem.appendChild(formEl);
+
+  // 확인버튼 클릭 시 입력창에 입력한 내용으로 수정.
+  formEl.onsubmit = (e) => {
+    e.preventDefault; //페이지 리로드 방지
+    const updatedTitle = inputEl.value;
+    const updatedTodo = {
+      title: updatedTitle,
+    };
+
+    fetch(API_URL + "/" + todoId, {
+      method: "PATCH",
+      headers: {
+        "Content-Type": "application/json",
+      },
+      body: JSON.stringify(updatedTodo),
+    })
+      .then((response) => response.json())
+      .then(() => {
+        return fetch(API_URL);
+      })
+      .then((response) => response.json())
+      .then((data) => renderTodo(data));
+  };
+
+  //취소 버튼 클릭시 원래 내용으로 복원
+  cancelBtn.onclick = () => {
+    todoItem.textContent = originalTitle;
+
+    const deleteEl = document.createElement("span");
+    deleteEl.textContent = "🗑️";
+    deleteEl.className = "deleteBtn";
+    deleteEl.onclick = () => deleteTodo(todoId);
+
+    const updateEl = document.createElement("span");
+    updateEl.textContent = "✏️";
+    updateEl.onclick = () => updateTodo(todoId, originalTitle);
+
+    todoItem.appendChild(deleteEl);
+    todoItem.appendChild(updateEl);
+  };
+};
```

Comment:
- 간단하게는 이미 renderTodo가 만들어져 있으니까 API 한 번 더 호출해서 새로 화면에 그리는 방법도 생각나네요!
- API 호출 없이 이미 front에 받아온 정보만 사용하려면 todoItem 이라는 element를 이미 만들었으니 이를 미리 분리해놓고 다시 사용하는 방법도 가능해 보입니다!

```suggestion
  cancelBtn.onclick = () => {
    fetch(API_URL)
      .then((response) => response.json())
      .then((data) => renderTodo(data));
  };
};
```

Reply:
감사합니다. 관련 내용 수정했습니다.